File: front/lib/api/src/models/extra.dart
import 'package:blee/api/src/models/image.dart';
import 'package:freezed_annotation/freezed_annotation.dart';

part 'extra.freezed.dart';
part 'extra.g.dart';

@freezed
class Extra with _$Extra {
  const factory Extra({
    required String id,
    required String name,
    required Image? thumbnail,
    @JsonKey(name: "package_id") required String? packageId,
    @JsonKey(name: "artist_id") required String artistId,
    @JsonKey(name: "artist_name") required String artistName,
    @JsonKey(name: "file_id") required String fileId,
    required int? discIndex,
    required int duration,
    required int? trackIndex,
    required List<ExtraType> type,
  }) = _Extra;

  factory Extra.fromJson(Map<String, dynamic> json) => _$ExtraFromJson(json);
}

enum ExtraType {
  @JsonValue("alternate_view")
  aternateView,
  @JsonValue("backdrops")
  backdrops,
  @JsonValue("behind_the_scenes")
  behindTheScenes,
  @JsonValue("interview")
  interview,
  @JsonValue("music_video")
  musicVideo,
  @JsonValue("other")
  other,
  @JsonValue("performance")
  performance,
  @JsonValue("trailer")
  trailer,
}

File: front-vue/models/domain/extra.ts
import * as yup from "yup";
import { Resource } from "./resource";
import { Image } from "./image";

const ExtraTypes = [
    "alternate_view",
    "backdrops",
    "behind_the_scenes",
    "interview",
    "music_video",
    "other",
    "performance",
    "trailer",
] as const;

type ExtraType = (typeof ExtraTypes)[number];

const Extra = Resource.concat(
    yup.object({
        name: yup.string().required(),
        thumbnail: Image.required().nullable(),
        package_id: yup.string().required(),
        artist_id: yup.string().required(),
        artist_name: yup.string().required(),
        file_id: yup.string().required(),
        disc_index: yup.number().required().nullable(),
        track_index: yup.number().required().nullable(),
        duration: yup.number().required(),
        type: yup
            .array(yup.mixed<ExtraType>().oneOf(ExtraTypes).required())
            .required(),
    }),
);

type Extra = yup.InferType<typeof Extra>;

type ExtraSortingKeys =
    | "name"
    | "artistName"
    | "packageName"
    | "addDate"
    | "releaseDate";

export { type ExtraType, type ExtraSortingKeys, Extra };
